examples/medplum-chart-demo/src/bots/core/charting-utils.ts

```diff
@@ -0,0 +1,366 @@
+    conditions.push({
+      ...partialCondition,
+      resourceType: 'Condition',
+      subject: encounter.subject as Reference<Patient>,
+      encounter: { reference: getReferenceString(encounter) },
+      recorder: { reference: getReferenceString(user) },
+      asserter: { reference: getReferenceString(user) },
+      category: [
+        {
+          coding: [
+            {
+              system: 'http://hl7.org/fhir/ValueSet/condition-category',
+              code: 'problem-list-item',
+              display: 'Problem List Item',
+            },
+          ],
+        },
+      ],
+    });
+  }
+
+  return conditions;
+}
+
+/**
+ * This function takes ClinicalImpression data and creates a ClinicalImpression resource. The ClinicalImpression resource
+ * represents any notes on an encounter in this context.
+ *
+ * @param encounter - The encounter that the data is derived from
+ * @param user - The user creating the ClinicalImpressions
+ * @param note - A string of any notes that a Practitioner may have taken during the encounter
+ * @returns A ClinicalImpression resource
+ */
+export function createClinicalImpressions(
+  encounter: Encounter,
+  user: Practitioner,
+  note?: string
+): ClinicalImpression | undefined {
+  if (!note) {
+    return undefined;
+  }
+
+  // Create the clinical impression
+  const clinicalImpression: ClinicalImpression = {
+    resourceType: 'ClinicalImpression',
+    status: 'in-progress',
+    subject: encounter.subject as Reference<Patient>,
+    encounter: { reference: getReferenceString(encounter) },
+    assessor: { reference: getReferenceString(user) },
+    note: [{ text: note }],
+  };
+
+  // Return the clinical impressions
+  return clinicalImpression;
+}
+
+/**
+ * This function takes an array of resources and creates a bundle that can be executed to create multiple resources at
+ * once. It will perform an upsert on the resources so that they are not duplicated.
+ *
+ * @param resources - An array of Conditions, Observations, and ClinicalImpressions to be added to the bundle
+ * @returns A bundle entry that can be executed to simultaneously create all the necessary Conditions, Observations, and
+ * ClinicalImpressions
+ */
+export function createBundle(resources: (Condition | Observation | ClinicalImpression)[]): Bundle {
+  const bundle: Bundle = {
+    resourceType: 'Bundle',
+    type: 'transaction',
+  };
+
+  const entries: BundleEntry[] = resources.map((resource) => {
+    const entry: BundleEntry = {
+      fullUrl: `urn:uuid:${randomUUID()}`,
+      request: { method: 'PUT', url: getUpsertUrl(resource) },
+      resource,
+    };
+
+    return entry;
+  });
+
+  bundle.entry = entries;
+
+  return bundle;
+}
+
+/**
+ *
+ * @param resource - The resource that is being upserted
+ * @returns A search url that can be used to upsert a resource
+ */
+function getUpsertUrl(resource: Resource): string {
+  if (resource.resourceType === 'Observation') {
+    const code = resource.code;
```

Comment:
To simplify, just always enforce that `Observations`  have a `Observation.code.coding[0].code` field. The you can always use the code in the url